File: ios/Sources/DKMadsSSPSDK/UI/DKMadsInterstitialAd.swift
import Foundation
import UIKit

@objc public protocol DKMadsInterstitialAdDelegate: AnyObject {
    @objc optional func interstitialAdDidReceiveAd(_ ad: DKMadsInterstitialAd)
    @objc optional func interstitialAd(_ ad: DKMadsInterstitialAd, didFailToReceiveAdWithError error: Error)
    @objc optional func interstitialAdDidPresent(_ ad: DKMadsInterstitialAd)
    @objc optional func interstitialAdDidDismiss(_ ad: DKMadsInterstitialAd)
}

/// Fullscreen interstitial (video, image, HTML5, or tag). Use for dashboard format `interstitial`.
@objc(DKMadsInterstitialAd)
public final class DKMadsInterstitialAd: NSObject, DKMadsFullScreenPresenting {
    @objc public weak var delegate: DKMadsInterstitialAdDelegate?
    @objc public weak var fullScreenContentDelegate: DKMadsFullScreenContentDelegate?
    @objc public let adUnitID: String
    @objc public private(set) var responseInfo: DKMadsResponseInfo?
    @objc public private(set) var loadedAd: Ad?
    private var loadedAt: Date?

    private var presenter: DKMadsInterstitialPresenter?

    private var lastAdRequest: DKMadsAdRequest?

    @objc public init(adUnitID: String) {
        self.adUnitID = adUnitID
    }

    /// Loads an interstitial ad. Swift API (`load` avoids ObjC `+load` selector clash).
    public static func load(
        adUnitID: String,
        adSize: CGSize? = nil,
        request: DKMadsAdRequest? = nil,
        completion: @escaping (DKMadsInterstitialAd?, Error?) -> Void
    ) {
        DKMadsInterstitialAd(adUnitID: adUnitID).load(adSize: adSize, request: request, completion: completion)
    }

    /// ObjC entry point — do not name `load` (reserved `+load` on `NSObject`).
    @objc(loadInterstitialWithAdUnitID:request:completion:)
    public static func loadInterstitial(
        adUnitID: String,
        request: DKMadsAdRequest?,
        completion: @escaping (DKMadsInterstitialAd?, Error?) -> Void
    ) {
        load(adUnitID: adUnitID, adSize: nil, request: request, completion: completion)
    }

    @objc(loadInterstitialWithAdUnitID:adWidth:adHeight:request:completion:)
    public static func loadInterstitial(
        adUnitID: String,
        adWidth: CGFloat,
        adHeight: CGFloat,
        request: DKMadsAdRequest?,
        completion: @escaping (DKMadsInterstitialAd?, Error?) -> Void
    ) {
        let size = (adWidth > 0 && adHeight > 0) ? CGSize(width: adWidth, height: adHeight) : nil
        load(adUnitID: adUnitID, adSize: size, request: request, completion: completion)
    }

    public func load(
        adSize: CGSize? = nil,
        request: DKMadsAdRequest? = nil,
        completion: @escaping (DKMadsInterstitialAd?, Error?) -> Void
    ) {
        if let request { lastAdRequest = request }
        let effectiveRequest = request ?? lastAdRequest
        guard SSPSDK.shared.isSDKInitialized else {
            completion(nil, DKMadsAdError.notInitialized.nsError())
            return
        }
        let bidSizes = Self.bidSizes(adUnitID: adUnitID, adSize: adSize)
        SSPSDK.shared.loadAd(
            code: adUnitID,
            format: .interstitial,
            sizes: bidSizes,
            placementCode: effectiveRequest?.placementCode,
            placementContext: effectiveRequest?.placementContext,
            keyValues: effectiveRequest?.keyValues ?? [:]
        ) { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let response):
                self.responseInfo = response.responseInfo
                guard response.success, let ad = response.ad, ad.hasFill else {
                    let err = DKMadsAdError.noFill.nsError(userInfo: [
                        NSLocalizedDescriptionKey: response.reason ?? "no_fill",
                    ])
                    self.delegate?.interstitialAd?(self, didFailToReceiveAdWithError: err)
                    completion(nil, err)
                    return
                }
                self.loadedAd = ad
                self.loadedAt = Date()
                self.delegate?.interstitialAdDidReceiveAd?(self)
                completion(self, nil)
            case .failure(let error):
                self.delegate?.interstitialAd?(self, didFailToReceiveAdWithError: error)
                completion(nil, error)
            }
        }
    }

    /// IAB interstitial tokens for bid matching — not raw UIScreen pixel dimensions.
    static func bidSizes(adUnitID: String, adSize: CGSize?) -> [CGSize] {
        if let adSize, adSize.width > 0, adSize.height > 0 {
            return [adSize]
        }
        let registered = SSPSDK.shared.registeredSizes(for: adUnitID)
        if !registered.isEmpty {
            return registered
        }
        return [CGSize(width: 320, height: 480)]
    }

    @objc(presentFromRootViewController:)
    public func present(from rootViewController: UIViewController) {
        guard let loadedAd else {
            delegate?.interstitialAd?(self, didFailToReceiveAdWithError: DKMadsAdError.noFill.nsError())
            return
        }
        if DKMadsAdCachePolicy.isExpired(loadedAt: loadedAt, format: .interstitial) {
            let err = DKMadsAdError.adExpired.nsError()
            delegate?.interstitialAd?(self, didFailToReceiveAdWithError: err)
            fullScreenContentDelegate?.ad?(self, didFailToPresentFullScreenContentWithError: err)
            return
        }
        let vc = DKMadsInterstitialPresenter(adUnitID: adUnitID, ad: loadedAd)
        vc.onDismiss = { [weak self] in
            guard let self else { return }
            self.presenter = nil
            self.delegate?.interstitialAdDidDismiss?(self)
            self.fullScreenContentDelegate?.adDidDismissFullScreenContent?(self)
        }
        vc.onPlaybackComplete = { [weak self] in
            self?.presenter?.dismiss(animated: true)
        }
        vc.onRenderFailed = { [weak self] error in
            guard let self else { return }
            self.presenter = nil
            rootViewController.dismiss(animated: true)
            self.delegate?.interstitialAd?(self, didFailToReceiveAdWithError: error)
            self.fullScreenContentDelegate?.ad?(self, didFailToPresentFullScreenContentWithError: error)
        }
        presenter = vc
        fullScreenContentDelegate?.adWillPresentFullScreenContent?(self)
        rootViewController.present(vc, animated: true) { [weak self] in
            guard let self else { return }
            self.delegate?.interstitialAdDidPresent?(self)
        }
    }
}

File: android/DKMadsInterstitialAd.kt
package com.dkmads.ssp

import android.app.Activity
import android.content.Context
import android.content.Intent
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Fullscreen interstitial (video, image, HTML5, or tag). Use when the dashboard ad unit format is `interstitial`.
 *
 * Bid sizes use explicit `adWidth`×`adHeight`, then [SSPSDK.registeredSizes], then **320×480** (not display pixels).
 */
class DKMadsInterstitialAd(
    override val adUnitId: String,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Main + SupervisorJob()),
) : DKMadsFullScreenPresenting {
    interface Listener {
        fun onAdLoaded(interstitial: DKMadsInterstitialAd, ad: Ad, responseInfo: DKMadsResponseInfo) {}
        fun onAdFailed(interstitial: DKMadsInterstitialAd, message: String, responseInfo: DKMadsResponseInfo?) {}
        fun onAdPresented(interstitial: DKMadsInterstitialAd) {}
        fun onAdDismissed(interstitial: DKMadsInterstitialAd) {}
    }

    var listener: Listener? = null
    var fullScreenContentCallback: DKMadsFullScreenContentCallback? = null
    var loadedAd: Ad? = null
        private set
    private var loadedAt: java.util.Date? = null
    override var responseInfo: DKMadsResponseInfo? = null
        private set

    var adWidth: Int = 0
    var adHeight: Int = 0

    fun load(
        context: Context,
        placementCode: String? = null,
        placementContext: String? = null,
        keyValues: Map<String, Any> = emptyMap(),
    ) {
        if (adUnitId.isBlank()) {
            listener?.onAdFailed(this, "adUnitId is required", null)
            return
        }
        scope.launch {
            val sizes = bidSizes(adUnitId, adWidth, adHeight)
            val result = SSPSDK.loadAd(
                context = context,
                adUnitCode = adUnitId,
                format = AdFormat.INTERSTITIAL,
                sizes = sizes,
                placementCode = placementCode,
                placementContext = placementContext,
                keyValues = keyValues,
            )
            result.fold(
                onSuccess = { ad ->
                    val info = DKMadsResponseInfo.from(ad)
                    responseInfo = info
                    if (!ad.hasFill) {
                        val msg = ad.reason ?: "no_fill"
                        listener?.onAdFailed(this@DKMadsInterstitialAd, msg, info)
                        return@fold
                    }
                    loadedAd = ad
                    loadedAt = java.util.Date()
                    listener?.onAdLoaded(this@DKMadsInterstitialAd, ad, info)
                },
                onFailure = { err ->
                    listener?.onAdFailed(this@DKMadsInterstitialAd, err.message ?: "load failed", null)
                },
            )
        }
    }

    /** Presents a fullscreen activity. Call from an [Activity] context when possible. */
    fun show(context: Context) {
        val ad = loadedAd
        if (ad == null || !ad.hasFill) {
            listener?.onAdFailed(this, "no_fill", responseInfo)
            return
        }
        if (DKMadsAdCachePolicy.isExpired(loadedAt, AdFormat.INTERSTITIAL)) {
            listener?.onAdFailed(this, "ad_expired", responseInfo)
            fullScreenContentCallback?.onAdFailedToShowFullScreenContent("ad_expired")
            return
        }
        DKMadsInterstitialActivity.present(
            context = context,
            adUnitId = adUnitId,
            ad = ad,
            callbacks = DKMadsInterstitialActivity.Callbacks(
                onPresented = {
                    listener?.onAdPresented(this@DKMadsInterstitialAd)
                    fullScreenContentCallback?.onAdShowedFullScreenContent()
                },
                onDismissed = {
                    listener?.onAdDismissed(this@DKMadsInterstitialAd)
                    fullScreenContentCallback?.onAdDismissedFullScreenContent()
                },
                onRenderFailed = { msg ->
                    listener?.onAdFailed(this@DKMadsInterstitialAd, msg, responseInfo)
                    fullScreenContentCallback?.onAdFailedToShowFullScreenContent(msg)
                },
            ),
        )
    }

    fun destroy() {
        scope.cancel()
        loadedAd = null
        responseInfo = null
    }

    companion object {
        fun load(
            context: Context,
            adUnitId: String,
            adWidth: Int = 0,
            adHeight: Int = 0,
            placementCode: String? = null,
            placementContext: String? = null,
            keyValues: Map<String, Any> = emptyMap(),
            onComplete: (DKMadsInterstitialAd?, String?) -> Unit,
        ) {
            val interstitial = DKMadsInterstitialAd(adUnitId).apply {
                this.adWidth = adWidth
                this.adHeight = adHeight
                listener = object : Listener {
                    override fun onAdLoaded(interstitial: DKMadsInterstitialAd, ad: Ad, responseInfo: DKMadsResponseInfo) {
                        onComplete(interstitial, null)
                    }

                    override fun onAdFailed(interstitial: DKMadsInterstitialAd, message: String, responseInfo: DKMadsResponseInfo?) {
                        onComplete(null, message)
                    }
                }
            }
            interstitial.load(context, placementCode, placementContext, keyValues)
        }

        /** IAB interstitial tokens for bid matching — not raw display pixel dimensions. */
        internal fun bidSizes(adUnitId: String, width: Int, height: Int): List<Pair<Int, Int>> {
            if (width > 0 && height > 0) return listOf(width to height)
            val registered = SSPSDK.registeredSizes(adUnitId)
            if (registered.isNotEmpty()) return registered
            return listOf(320 to 480)
        }
    }
}
